crypto/dso/dso_dlfcn.c

```diff
@@ -307,6 +307,71 @@ static int dladdr(void *address, Dl_info *dl)
     return (int)v;
 }
 # endif                         /* __sgi */
+# ifdef _AIX
+/*-
+ * See IBM's AIX Version 7.2, Technical Reference:
+ *  Base Operating System and Extensions, Volume 1 and 2
+ *  https://www.ibm.com/support/knowledgecenter/ssw_aix_72/com.ibm.aix.base/technicalreferences.htm
+ */
+#  include <sys/ldr.h>
+#  include <errno.h>
+/* ~ 64 * (sizeof(struct ld_info) + _XOPEN_PATH_MAX + _XOPEN_NAME_MAX) */
+#  define DLFCN_LDINFO_SIZE 86976
+extern int errno;
+typedef struct Dl_info {
+    const char *dli_fname;
+} Dl_info;
+/* NOTE: when looking up a function symbol, the caller of DSO_...byaddr() /
+ *  OPENSSL_atexit() might need to consider the Pointer Glue virtual address
+ *  handling on AIX/PowerPC (aka De-Virtualization; related to the ptrgl
+ *  compiler flags). The real address of a function can be retrieved with:
+ *  (void*)*((ulong*)function)
+ *  This dladdr()-implementation will also find the ptrgl virtual address of a
+ *  function, which is just located in the DATA segment instead of the TEXT
+ *  segment. */
+static int dladdr(void *address, Dl_info *dl)
+{
+    unsigned int found = 0;
+    struct ld_info *ldinfos, *next_ldi, *this_ldi;
+
+    if ((ldinfos = (struct ld_info *)OPENSSL_malloc(DLFCN_LDINFO_SIZE)) == NULL) {
+      errno = ENOMEM;
+      dl->dli_fname = NULL;
+      return 0;
+    }
+
+    if ((loadquery(L_GETINFO, (void *)ldinfos, DLFCN_LDINFO_SIZE)) < 0) {
+      /* Error handling is done through errno and dlerror() reading errno:
+       *  ENOMEM (ldinfos buffer is too small),
+       *  EINVAL (invalid flags),
+       *  EFAULT (invalid ldinfos ptr)
+       */
+      OPENSSL_free((void *)ldinfos);
+      dl->dli_fname = NULL;
+      return 0;
+    }
+    next_ldi = ldinfos;
+
+    do {
+      this_ldi = next_ldi;
+      if (((address >= this_ldi->ldinfo_textorg) &&
+           (address < (this_ldi->ldinfo_textorg + this_ldi->ldinfo_textsize))) ||
+          ((address >= this_ldi->ldinfo_dataorg) &&
+           (address < (this_ldi->ldinfo_dataorg + this_ldi->ldinfo_datasize)))) {
+        found = 1;
+        /* Ignoring the possibility of a member name and just returning the
+         * path name. See docs: ldr.h, loadquery() and dlopen()/RTLD_MEMBER. */
+        if ((dl->dli_fname = OPENSSL_strdup(this_ldi->ldinfo_filename)) == NULL) {
+          errno = ENOMEM;
+        }
+      } else {
+        next_ldi = (void *)this_ldi + this_ldi->ldinfo_next;
```

Comment:
Wouldn't compiler get grumpy about pointer arithmetic with (void *)?

Reply:
No it is needed and at least the xlC 13.1.3 doesn't give any hint there might be something dubious. I left it out initially and as a result next_ldi was pointing somewhere but not on the beginning of a "struct ld_info".

Reply:
I rather meant that it should be `(char *)`. xlc might be ok with, but gcc can issue warning. Well, maybe not by default, but it will if -Wpointer-arith slips through. And we kind of frown on warnings...